os/drivers/mipidsi/mipi_dsi_device.c

```diff
@@ -0,0 +1,950 @@
+		if (tx == NULL) {
+			return -ENOMEM;
+		}
+	}
+
+	/* concatenate the DCS command byte and the payload */
+
+	tx[0] = cmd;
+	if (data != NULL) {
+		memcpy(&tx[1], data, len);
+	}
+
+	ret = mipi_dsi_dcs_write_buffer(device, tx, len + 1);
+
+	if (tx != stack_tx) {
+		kmm_free(tx);
+	}
+
+	return ret;
+}
+
+/****************************************************************************
+ * Name: mipi_dsi_dcs_read
+ *
+ * Description:
+ *   Send DCS read request command.
+ *
+ * Input Parameters:
+ *   device - DSI peripheral device
+ *   cmd - DCS command
+ *   data - buffer in which to receive data
+ *   len - size of receive buffer
+ *
+ * Returned Value:
+ *   The number of bytes read or a negative error code on failure.
+ *
+ ****************************************************************************/
+
+ssize_t mipi_dsi_dcs_read(FAR struct mipi_dsi_device *device, uint8_t cmd, FAR void *data, size_t len)
+{
+	struct mipi_dsi_msg msg;
+
+	msg.channel = device->channel;
+	msg.type = MIPI_DSI_DCS_READ_0_PARAM;
+	msg.tx_buf = &cmd;
+	msg.tx_len = 1;
+	msg.rx_buf = data;
+	msg.rx_len = len;
+	msg.flags = 0;
+
+	return mipi_dsi_transfer(device, &msg);
+}
+
+/****************************************************************************
+ * Name: mipi_dsi_dcs_nop
+ *
+ * Description:
+ *   Send DCS nop packet
+ *
+ * Input Parameters:
+ *   device - DSI peripheral device
+ *
+ * Returned Value:
+ *   OK on success or a negative error code on failure.
+ *
+ ****************************************************************************/
+
+int mipi_dsi_dcs_nop(FAR struct mipi_dsi_device *device)
+{
+	int ret;
+
+	ret = mipi_dsi_dcs_write(device, MIPI_DCS_NOP, NULL, 0);
+	return ret < 0 ? ret : OK;
+}
+
+/****************************************************************************
+ * Name: mipi_dsi_dcs_soft_reset
+ *
+ * Description:
+ *   Send a software reset of the display module
+ *
+ * Input Parameters:
+ *   device - DSI peripheral device
+ *
+ * Returned Value:
+ *   OK on success or a negative error code on failure.
+ *
+ ****************************************************************************/
+
+int mipi_dsi_dcs_soft_reset(FAR struct mipi_dsi_device *device)
+{
+	int ret;
+
+	ret = mipi_dsi_dcs_write(device, MIPI_DCS_SOFT_RESET, NULL, 0);
+	return ret < 0 ? ret : OK;
+}
+
+/****************************************************************************
+ * Name: mipi_dsi_dcs_get_power_mode
+ *
+ * Description:
+ *   Query the display module's current power mode
+ *
+ * Input Parameters:
+ *   device - DSI peripheral device
+ *   mode - return location of the current power mode
+ *
+ * Returned Value:
+ *   OK on success or a negative error code on failure.
+ *
+ ****************************************************************************/
+
+int mipi_dsi_dcs_get_power_mode(FAR struct mipi_dsi_device *device, FAR uint8_t *mode)
+{
+	int ret;
+
+	ret = mipi_dsi_dcs_read(device, MIPI_DCS_GET_POWER_MODE, mode, sizeof(*mode));
+	if (ret <= 0) {
```

Comment:
```
	if (ret < 0) {
                return ret;
        }
	else if (ret == 0) {
		ret = -ENODATA;
	}
        else {
		return OK;
	}
```

Reply:
ok, will change this. 
For the else part, Mr. chang gave in the previous  comment to align like this ---> } else {